File: together-sandbox-typescript/src/Snapshots.ts
import * as fs from "fs";
import * as os from "os";
import * as path from "path";
import * as api from "./api-clients/api/index.js";
import { type Client as ApiClient } from "./api-clients/api/client/index.js";
import { isLocalEnvironment } from "./configuration.js";
import { callApi, withRetry } from "./utils.js";
import { Page } from "./pagination.js";
import {
  buildDockerImage,
  dockerLogin,
  isDockerAvailable,
  pushDockerImage,
} from "./docker.js";
import { RemoteImageBuilderClient } from "./RemoteImageBuilder.js";
import { randomUUID } from "crypto";
import type { RetryConfig } from "./types.js";

export type SnapshotProgress = { output: string } & (
  | { step: "prepare" }
  | { step: "build" }
  | { step: "auth" }
  | { step: "push" }
  | { step: "register" }
  | { step: "alias" }
);

export type Snapshot = api.Snapshot;

/**
 * Parameters for creating a snapshot
 */
export type CreateContextSnapshotParams = {
  context: string;
  dockerfile?: string;
  alias?: string;
  onProgress?: (event: SnapshotProgress) => void;
  /** Arbitrary key/value labels to attach to the snapshot. */
  tags?: Record<string, string>;
  /** Seconds after creation before the snapshot is automatically retired. */
  ttl?: number;
};

export type CreateImageSnapshotParams = {
  image: string;
  alias?: string;
  onProgress?: (event: SnapshotProgress) => void;
  /** Arbitrary key/value labels to attach to the snapshot. */
  tags?: Record<string, string>;
  /** Seconds after creation before the snapshot is automatically retired. */
  ttl?: number;
};
export type CreateSnapshotParams =
  | CreateContextSnapshotParams
  | CreateImageSnapshotParams;

/**
 * Result of a successful snapshot creation.
 */
export interface CreateSnapshotResult {
  /** ID of the created snapshot. */
  snapshotId: string;
  /** The alias that was applied, if any. */
  alias?: string;
}

function stripAnsiCodes(str: string) {
  // Matches ESC [ params … finalChar
  //   \x1B       = ESC
  //   \[         = literal "["
  //   [0-?]*     = any parameter bytes (digits, ;, ?)
  //   [ -/]*     = any intermediate bytes (space through /)
  //   [@-~]      = final byte ( @ A–Z [ \ ] ^ _ ` a–z { | } ~ )
  const CSI_REGEX = /\x1B\[[0-?]*[ -/]*[@-~]/g;
  return str.replace(CSI_REGEX, "");
}

export type ImageReference = {
  registry?: string;
  repository?: string;
  name: string;
  tag?: string;
};

/**
 * Snapshot build and management operations, accessed as `sdk.snapshots.*`.
 */
export class SnapshotsNamespace {
  constructor(
    private readonly _apiClient: ApiClient,
    private readonly _baseUrl: string,
    private readonly _apiKey: string,
    private readonly _retryConfig?: RetryConfig,
  ) {}

  // ─── Public entry points ──────────────────────────────────────────────────

  async getById(id: string): Promise<Snapshot> {
    const result = await callApi(
      "api.snapshots.getById",
      () =>
        api.getSnapshot({
          client: this._apiClient,
          path: { id },
        }),
      this._retryConfig,
    );

    return result;
  }

  async getByAlias(alias: string): Promise<Snapshot> {
    const result = await callApi(
      "api.snapshots.getByAlias",
      () =>
        api.getSnapshotByAlias({
          client: this._apiClient,
          path: { alias },
        }),
      this._retryConfig,
    );

    return result;
  }

  /**
   * List snapshots.
   *
   * Returns a {@link Page} that is async-iterable across all pages — iterate it
   * directly to walk every snapshot, or use `getNextPage()` / `nextCursor` for
   * manual page-by-page control.
   *
   * @param options.limit Max items per page (1–100, default 20).
   * @param options.cursor Resume from a cursor returned by a previous page's
   *   {@link Page.nextCursor} (omit to start from the first page).
   * @param options.excludeRetired When true, retired snapshots are excluded
   *   (default false — retired snapshots are included).
   * @param options.tags Filter by tags; matches snapshots whose tags contain
   *   all the given pairs.
   */
  async list(options?: {
    limit?: number;
    cursor?: string;
    excludeRetired?: boolean;
    tags?: Record<string, string>;
  }): Promise<Page<Snapshot>> {
    const fetchPage = async (cursor?: string): Promise<Page<Snapshot>> => {
      const result = await callApi(
        "api.snapshots.list",
        () =>
          api.listSnapshots({
            client: this._apiClient,
            query: {
              limit: options?.limit,
              cursor,
              exclude_retired: options?.excludeRetired,
              tags: options?.tags,
            },
          }),
        this._retryConfig,
      );
      return new Page<Snapshot>(result.data, result.next_cursor, fetchPage);
    };

    return fetchPage(options?.cursor);
  }

  async alias(snapshotId: string, alias: string): Promise<void> {
    await callApi(
      "api.snapshots.alias",
      () =>
        api.aliasSnapshot({
          client: this._apiClient,
          path: { snapshot_id: snapshotId },
          body: { alias },
        }),
      this._retryConfig,
    );
  }

  /**
   * Retire a snapshot by id and return the retired snapshot.
   *
   * Once retired, the snapshot can no longer be used to create new sandboxes,
   * and it is eventually deleted, but only once no sandbox still references
   * it.
   */
  async retire(id: string): Promise<Snapshot> {
    const result = await callApi(
      "api.snapshots.retire",
      () =>
        api.retireSnapshot({
          client: this._apiClient,
          path: { id },
        }),
      this._retryConfig,
    );

    return result;
  }

  /**
   * Create a snapshot from a Docker build context or a public Docker image.
   *
   * Pass `{ context, dockerfile?, alias?, onProgress? }` to build from a Dockerfile.
   * Pass `{ image, alias?, onProgress? }` to register a public Docker image.
   *
   * @note The `snapshots.create` operation is not idempotent: retrying on a 500 error
   * after the snapshot has been created will register a duplicate. If you are using
   * retry configuration, consider excluding this operation via:
   * `shouldRetry: ({ operation }) => operation !== 'snapshots.create'`
   */
  async create(params: CreateSnapshotParams): Promise<CreateSnapshotResult> {
    // Build from context — defaults to the remote image-builder service.
    // Set `TOGETHER_LOCAL_BUILD=1` to fall back to the legacy local Docker
    // build + push flow (`_buildAndRegister`), which mirrors the Python SDK.
    let result: { image: string; architecture: "amd64" | "arm64" };
    if ("context" in params && process.env.TOGETHER_LOCAL_BUILD === "1") {
      const dockerAvailable = await isDockerAvailable();
      if (!dockerAvailable) {
        throw new Error(
          "Docker is not available. Please install Docker to use local build mode (TOGETHER_LOCAL_BUILD=1).",
        );
      }
      result = await this._buildAndRegister(params);
    } else if ("context" in params) {
      result = await this._buildRemotely(params);
    } else {
      result = await this._buildRemotelyFromImage(params);
    }

    params.onProgress?.({
      step: "register",
      output: "Registering snapshot...",
    });
    const snapshotData = await callApi(
      "api.snapshots.create",
      () =>
        api.createSnapshot({
          client: this._apiClient,
          body: {
            image: result.image,
            architecture: result.architecture,
            ttl: params.ttl,
            tags: params.tags,
          },
        }),
      this._retryConfig,
    );

    const snapshotId = snapshotData.id;

    // Create alias if needed
    if (params.alias) {
      const alias = params.alias;

      params.onProgress?.({
        step: "alias",
        output: "Creating alias...",
      });

      await callApi(
        "api.snapshots.alias",
        () =>
          api.aliasSnapshot({
            client: this._apiClient,
            path: { snapshot_id: snapshotId },
            body: { alias },
          }),
        this._retryConfig,
      );
    }

    return {
      snapshotId,
      alias: params.alias,
    };
  }

  // ─── Private helpers ──────────────────────────────────────────────────────
  /**
   * Build a Docker image from a public image reference via the remote
   * image-builder service.
   *
   * Creates a temporary directory named after the image (so `_buildRemotely`
   * derives a meaningful image name from `path.basename(contextDir)`) and
   * writes a single-line Dockerfile of the form `FROM <image>` into it. The
   * temp directory is removed on exit regardless of build outcome.
   */
  private async _buildRemotelyFromImage(
    params: CreateImageSnapshotParams,
  ): Promise<{ image: string; architecture: "amd64" | "arm64" }> {
    // Derive a folder name from the image's last path component, stripped of
    // any tag/digest. e.g. "ghcr.io/foo/bar:latest" -> "bar",
    // "node:22" -> "node". The fallback to "image" guards against
    // pathological inputs that strip to an empty string.
    const lastSegment = params.image.split("/").pop() ?? "";
    const imageNameSlug =
      lastSegment
        .split("@")[0]
        .split(":")[0]
        .toLowerCase()
        .replace(/_/g, "-") || "image";

    const tmpParent = await fs.promises.mkdtemp(
      path.join(os.tmpdir(), "together-sandbox-"),
    );
    const contextDir = path.join(tmpParent, imageNameSlug);
    try {
      await fs.promises.mkdir(contextDir, { recursive: true });
      await fs.promises.writeFile(
        path.join(contextDir, "Dockerfile"),
        `FROM ${params.image}\n`,
      );

      return await this._buildRemotely({
        context: contextDir,
        alias: params.alias,
        onProgress: params.onProgress,
      });
    } finally {
      await fs.promises.rm(tmpParent, { recursive: true, force: true });
    }
  }
  /**
   * Build a Docker image using the remote image-builder service.
   *
   * Mirrors the Python SDK's `_build_image_remotely`: derives the
   * image-builder URL from the configured base URL by replacing
   * `api.bartender.` with `builder.`, uses the SDK's API key as the
   * auth token, and returns the image reference and architecture that should
   * be passed to `snapshots.create`.
   */
  private async _buildRemotely(
    params: CreateContextSnapshotParams,
  ): Promise<{ image: string; architecture: "amd64" | "arm64" }> {
    const ibApiUrl = this._baseUrl.replace("api.bartender.", "builder.");

    const contextDir = path.resolve(params.context);
    const dockerfilePath = params.dockerfile
      ? path.resolve(params.dockerfile)
      : path.join(contextDir, "Dockerfile");
    const dockerfileRel = path.relative(contextDir, dockerfilePath);

    // Unique name/tag per build so concurrent builds can't collide on the
    // same ref (the namespace is shared, derived from the auth token).
    const imageName = `image-${randomUUID().toLowerCase()}`;
    const imageTag = randomUUID().toLowerCase();
    const imageRef = `${imageName}:${imageTag}`;

    const emit = (output: string) =>
      params.onProgress?.({ step: "build", output: stripAnsiCodes(output) });

    const ibClient = new RemoteImageBuilderClient({
      apiUrl: ibApiUrl,
      token: this._apiKey,
      logger: {
        debug: emit,
        warning: emit,
      },
    });

    params.onProgress?.({
      step: "prepare",
      output: "Starting remote build ...",
    });
    const imageRefStr = await ibClient.build({
      contextDir,
      imageName: imageRef,
      dockerfile: dockerfileRel,
      nydus: true,
    });

    const archEnv = process.env.TOGETHER_REMOTE_ARCHITECTURE;
    let architecture: "amd64" | "arm64";
    if (archEnv) {
      if (archEnv !== "amd64" && archEnv !== "arm64") {
        throw new Error(
          `Invalid TOGETHER_REMOTE_ARCHITECTURE=${JSON.stringify(archEnv)}; ` +
            `expected one of: ["amd64","arm64"]`,
        );
      }
      architecture = archEnv;
    } else {
      architecture = "amd64";
    }

    return {
      image: imageRefStr,
      architecture,
    };
  }

  private async _buildAndRegister(
    params: CreateContextSnapshotParams,
  ): Promise<{ image: string; architecture: "amd64" | "arm64" }> {
    const architecture: "amd64" | "arm64" =
      process.arch === "arm64" && isLocalEnvironment(this._baseUrl)
        ? "arm64"
        : "amd64";
    const dockerfilePath = params.dockerfile;
    const context = params.context;

    const credential = await callApi(
      "api.snapshots.issueContainerRegistryCredential",
      () =>
        api.issueContainerRegistryCredential({
          client: this._apiClient,
        }),
      this._retryConfig,
    );
    const registryUrl = credential.registry_url;
    const registryHost = registryUrl.split("/")[0];
    const imageName = `image-${randomUUID().toLowerCase()}`;
    const tag = randomUUID().toLowerCase();
    const fullImageName = `${registryUrl}/${imageName}:${tag}`;

    // Docker Build
    params.onProgress?.({
      step: "prepare",
      output: "Preparing snapshot...",
    });
    params.onProgress?.({ step: "build", output: "Building snapshot..." });
    await buildDockerImage({
      dockerfilePath,
      imageName: fullImageName,
      context,
      architecture,
      onOutput: (output: string) => {
        const cleanOutput = stripAnsiCodes(output);
        params.onProgress?.({ step: "build", output: cleanOutput });
      },
    });

    // Docker Login
    params.onProgress?.({ step: "auth", output: "Authenticating..." });
    await dockerLogin({
      registry: registryHost,
      username: credential.username,
      password: credential.password,
      onOutput: (output: string) => {
        const cleanOutput = stripAnsiCodes(output);
        params.onProgress?.({ step: "auth", output: cleanOutput });
      },
    });

    // Push Docker Image — wrapped in `withRetry` because `docker push` is
    // naturally idempotent (content-addressed layers) and transient registry
    // failures are common. We don't retry build (non-deterministic) or login
    // (fast, separate concern).
    params.onProgress?.({ step: "push", output: "Pushing Docker image..." });
    await withRetry(
      "snapshots.pushDockerImage",
      () =>
        pushDockerImage(fullImageName, (output: string) => {
          const cleanOutput = stripAnsiCodes(output);
          params.onProgress?.({ step: "push", output: cleanOutput });
        }),
      {
        ...this._retryConfig,
        onRetry: async (ctx) => {
          params.onProgress?.({
            step: "push",
            output: `Push failed (attempt ${ctx.attempt}), retrying in ${Math.round(ctx.delay)}ms…`,
          });
          await this._retryConfig?.onRetry?.(ctx);
        },
      },
    );

    return {
      image: fullImageName,
      architecture,
    };
  }
}

File: together-sandbox-python/together_sandbox/_snapshots.py
from __future__ import annotations


import asyncio
import os
import platform
import time

from collections.abc import Callable
from dataclasses import dataclass
from uuid import uuid4, UUID
from .api.client import AuthenticatedClient as ApiClient
from ._utils import (
    _strip_ansi,
    RetryConfig,
    RetryContext,
    _call_api,
    _with_retry,
)
from ._configuration import get_inferred_base_url, is_local_environment

# ── Snapshot API endpoint functions (detailed variants) ───────────────────────
from .api.api.default.create_snapshot import asyncio_detailed as create_snapshot_api
from .api.api.default.alias_snapshot import asyncio_detailed as alias_snapshot_api
from .api.api.default.get_snapshot_by_alias import (
    asyncio_detailed as get_snapshot_by_alias_api,
)
from .api.api.default.issue_container_registry_credential import (
    asyncio_detailed as issue_container_registry_credential_api,
)
from .api.api.default.get_snapshot import asyncio_detailed as get_snapshot_api
from .api.api.default.list_snapshots import asyncio_detailed as list_snapshots_api
from .api.api.default.delete_snapshot import asyncio_detailed as delete_snapshot_api
from .api.api.default.delete_snapshot_by_alias import (
    asyncio_detailed as delete_snapshot_by_alias_api,
)

# ── Snapshot API models ───────────────────────────────────────────────────────
from .api.models.alias_snapshot_body import AliasSnapshotBody
from .api.models.create_snapshot_body import CreateSnapshotBody
from .api.models.create_snapshot_body_architecture import CreateSnapshotBodyArchitecture
from .api.models.container_registry_credential import ContainerRegistryCredential
from .api.models.snapshot import Snapshot

# ── Helpers ────────────────────────────────────────────────────────────
from .docker import (
    DockerBuildOptions,
    DockerLoginOptions,
    build_docker_image,
    docker_login,
    is_docker_available,
    push_docker_image,
)

# ─── Remote Snapshot Builder ──────────────────────────────────────────────────────────
from ._remote_image_builder import RemoteImageBuilderClient

# ─── Snapshot types ──────────────────────────────────────────────────────────


@dataclass
class SnapshotProgress:
    step: str
    output: str


@dataclass
class CreateSnapshotResult:
    snapshot_id: str
    alias: str | None = None


@dataclass
class CreateContextSnapshotParams:
    context: str
    dockerfile: str | None = None
    alias: str | None = None
    on_progress: Callable[[SnapshotProgress], None] | None = None
    memory_snapshot: bool | None = None


@dataclass
class CreateImageSnapshotParams:
    image: str
    alias: str | None = None
    on_progress: Callable[[SnapshotProgress], None] | None = None


CreateSnapshotParams = CreateContextSnapshotParams | CreateImageSnapshotParams

# ─── SnapshotsNamespace ───────────────────────────────────────────────────────


class SnapshotsNamespace:
    """
    Snapshot build and management operations, accessed as ``sdk.snapshots.*``.
    """

    def __init__(
        self,
        api_client: ApiClient,
        base_url: str,
        *,
        retry: RetryConfig | None = None,
        api_key: str,
    ) -> None:
        self._api_client = api_client
        self._base_url = base_url
        self._retry = retry
        self._api_key = api_key

    # ─── Public entry points ──────────────────────────────────────────────────

    async def alias(self, snapshot_id: str, alias: str) -> None:
        """Create an alias for an existing snapshot"""
        await _call_api(
            "snapshots.alias",
            lambda: alias_snapshot_api(
                UUID(snapshot_id),
                client=self._api_client,
                body=AliasSnapshotBody(alias=alias),
            ),
            self._retry,
            context=f"for snapshot {snapshot_id!r}",
        )

    async def create(self, params: CreateSnapshotParams) -> CreateSnapshotResult:
        """Create a snapshot from either a Docker context or a public Docker image.

        .. note::
            The ``snapshots.create`` operation is not idempotent: retrying on a
            transient 500 error after the snapshot was already created will
            register a duplicate. If you use :class:`RetryConfig`, consider
            excluding this operation::

                RetryConfig(should_retry=lambda ctx: ctx.operation != "snapshots.create")
        """
        if isinstance(params, CreateContextSnapshotParams):
            # Context-based snapshot — requires Docker
            if not await is_docker_available():
                raise RuntimeError(
                    "Docker is not available. Please install Docker to use snapshot builds."
                )

            def _emit(step: str, output: str) -> None:
                if params.on_progress:
                    params.on_progress(SnapshotProgress(step=step, output=output))

            if os.getenv("TOGETHER_LOCAL_BUILD") == "1":
                result = await self._build_and_register(params)
            else:
                result = await self._build_image_via_builder(params)

            _emit("register", "Registering snapshot...")

            snapshot_data = await _call_api(
                "snapshots.create",
                lambda: create_snapshot_api(
                    client=self._api_client,
                    body=CreateSnapshotBody(
                        image=result["image"],
                        architecture=result["architecture"],
                    ),
                ),
                self._retry,
            )

            snapshot_id = str(snapshot_data.id)

            if params.alias:
                alias = params.alias
                _emit("alias", "Creating alias...")
                await _call_api(
                    "snapshots.alias",
                    lambda: alias_snapshot_api(
                        snapshot_data.id,
                        client=self._api_client,
                        body=AliasSnapshotBody(alias=alias),
                    ),
                    self._retry,
                )

            return CreateSnapshotResult(snapshot_id=snapshot_id, alias=params.alias)

        else:
            # Image-based snapshot — no Docker required
            def _emit(step: str, output: str) -> None:
                if params.on_progress:
                    params.on_progress(SnapshotProgress(step=step, output=output))

            _emit("register", "Creating snapshot from image...")
            snapshot_data = await _call_api(
                "snapshots.create",
                lambda: create_snapshot_api(
                    client=self._api_client,
                    body=CreateSnapshotBody(
                        image=params.image,
                        architecture=CreateSnapshotBodyArchitecture.AMD64,
                    ),
                ),
                self._retry,
            )

            snapshot_id = str(snapshot_data.id)

            if params.alias:
                alias = params.alias
                _emit("alias", "Creating alias...")
                await _call_api(
                    "snapshots.alias",
                    lambda: alias_snapshot_api(
                        snapshot_data.id,
                        client=self._api_client,
                        body=AliasSnapshotBody(alias=alias),
                    ),
                    self._retry,
                )

            return CreateSnapshotResult(snapshot_id=snapshot_id, alias=params.alias)

    async def get_by_id(self, id: str) -> Snapshot:
        """
        Get snapshot information by id.

        Args:
            id: Snapshot id

        Returns:
            Snapshot: Snapshot model with id, type, byte_size, and metadata

        Raises:
            RuntimeError: If the snapshot is not found or the API returns an
                application-level error response.
            httpx.TimeoutException: If the request to the snapshot API times out.

        Example:
            >>> snapshot = await sdk.snapshots.get("some-id")
            >>> print(snapshot.id)
            >>> print(snapshot.byte_size)
        """
        return await _call_api(
            "snapshots.getById",
            lambda: get_snapshot_api(UUID(id), client=self._api_client),
            self._retry,
            context=f"for id {id!r}",
        )

    async def get_by_alias(self, alias: str) -> Snapshot:
        """
        Get snapshot information by alias.

        Args:
            alias: Snapshot alias

        Returns:
            Snapshot: Snapshot model with id, type, byte_size, and metadata

        Raises:
            RuntimeError: If the snapshot is not found or API returns no data
            httpx.TimeoutException: If the API request fails

        Example:
            >>> snapshot = await sdk.snapshots.get_by_alias("my-app@latest")
            >>> print(snapshot.id)
            >>> print(snapshot.byte_size)
        """
        return await _call_api(
            "snapshots.getByAlias",
            lambda: get_snapshot_by_alias_api(
                alias,
                client=self._api_client,
            ),
            self._retry,
            context=f"for alias {alias!r}",
        )

    async def list(self) -> list[Snapshot]:
        """
        List snapshots.

        Returns:
            list[Snapshot]: List of snapshot models with id, type, byte_size, and metadata

        Raises:
            RuntimeError: If the API returns no data
            httpx.TimeoutException: If the API request fails

        Example:
            >>> snapshots = await sdk.snapshots.list()
            >>> for snapshot in snapshots:
            ...     print(snapshot.id)
        """
        return await _call_api(
            "snapshots.list",
            lambda: list_snapshots_api(client=self._api_client),
            self._retry,
        )

    async def delete_by_id(self, id: str) -> None:
        """
        Delete a snapshot by id.

        Args:
            id: Snapshot id

        Raises:
            RuntimeError: If the API request fails
        """
        await _call_api(
            "snapshots.deleteById",
            lambda: delete_snapshot_api(UUID(id), client=self._api_client),
            self._retry,
            context=f"for snapshot {id!r}",
        )

    async def delete_by_alias(self, alias: str) -> None:
        """
        Delete a snapshot by alias.

        Args:
            alias: Snapshot alias

        Raises:
            RuntimeError: If the API request fails
        """
        # Remove leading '@' if present (for consistency with API)
        clean_alias = alias.lstrip("@")

        await _call_api(
            "snapshots.deleteByAlias",
            lambda: delete_snapshot_by_alias_api(
                clean_alias,
                client=self._api_client,
            ),
            self._retry,
            context=f"for snapshot {alias!r}",
        )

    # ─── Private helpers ──────────────────────────────────────────────────────

    async def _build_image_via_builder(
        self,
        params: CreateContextSnapshotParams,
    ) -> dict[str, str | CreateSnapshotBodyArchitecture]:
        """
        Build a Docker image using the remote image-builder service.

        Derives the image-builder URL from the configured base URL by replacing
        "api.bartender." or "api." with "builder.". Uses the SDK's API key as
        the auth token. Returns a dict with ``image`` and ``architecture`` keys,
        compatible with ``create_snapshot``.

        Raises:
            RuntimeError: If the build fails.
        """
        from pathlib import Path

        csb_base_url = get_inferred_base_url()
        ib_api_url = csb_base_url.replace("api.bartender.", "builder.").replace(
            "api.", "builder."
        )

        context_dir = Path(os.path.realpath(params.context))
        dockerfile_path = (
            Path(os.path.realpath(params.dockerfile))
            if params.dockerfile
            else context_dir / "Dockerfile"
        )
        dockerfile_rel = str(dockerfile_path.relative_to(context_dir))

        image_name = context_dir.name.lower().replace("_", "-")
        image_tag = str(int(time.time()))

        # The server derives the namespace from the auth token; pass name:tag only.
        image_ref = f"{image_name}:{image_tag}"

        ib_client = RemoteImageBuilderClient(
            api_url=ib_api_url, token=self._api_key, logger=None
        )
        image_ref_str = await ib_client.build(
            context_dir=context_dir,
            image_name=image_ref,
            dockerfile=dockerfile_rel,
            nydus=True,
        )

        architecture_str = os.getenv("TOGETHER_REMOTE_ARCHITECTURE")
        if architecture_str:
            try:
                architecture = CreateSnapshotBodyArchitecture(architecture_str)
            except ValueError as e:
                raise RuntimeError(
                    f"Invalid TOGETHER_REMOTE_ARCHITECTURE={architecture_str!r}; "
                    f"expected one of: "
                    f"{[a.value for a in CreateSnapshotBodyArchitecture]}"
                ) from e
        else:
            architecture = CreateSnapshotBodyArchitecture.AMD64

        return {
            "image": image_ref_str,
            "architecture": architecture,
        }

    async def _build_and_register(
        self,
        params: CreateContextSnapshotParams,
    ) -> dict[str, str | CreateSnapshotBodyArchitecture]:
        architecture = (
            CreateSnapshotBodyArchitecture.ARM64
            if platform.machine().lower() == "arm64"
            and is_local_environment(self._base_url)
            else CreateSnapshotBodyArchitecture.AMD64
        )
        context = os.path.realpath(params.context)
        dockerfile_path = (
            os.path.realpath(params.dockerfile) if params.dockerfile else None
        )

        credential: ContainerRegistryCredential = await _call_api(
            "snapshots.issueContainerRegistryCredential",
            lambda: issue_container_registry_credential_api(client=self._api_client),
            self._retry,
        )
        registry_url = credential.registry_url
        registry_host = registry_url.split("/")[0]
        image_name = f"image-{uuid4()}".lower()
        image_tag = str(uuid4()).lower()
        full_image_name = f"{registry_url}/{image_name}:{image_tag}"

        def _emit(step: str, output: str) -> None:
            if params.on_progress:
                params.on_progress(SnapshotProgress(step=step, output=output))

        _emit("prepare", "Preparing snapshot...")
        _emit("build", "Building snapshot...")
        await build_docker_image(
            DockerBuildOptions(
                dockerfile_path=dockerfile_path,
                image_name=full_image_name,
                context=context,
                architecture=architecture,
                on_output=lambda out: _emit("build", _strip_ansi(out)),
            )
        )

        _emit("auth", "Authenticating...")
        await docker_login(
            DockerLoginOptions(
                registry=registry_host,
                username=credential.username,
                password=credential.password,
                on_output=lambda out: _emit("auth", _strip_ansi(out)),
            )
        )

        # Push Docker Image — wrapped in `_with_retry` because `docker push`
        # is naturally idempotent (content-addressed layers) and transient
        # registry failures are common. We don't retry build (non-deterministic)
        # or login (fast, separate concern).
        _emit("push", "Pushing Docker image...")

        async def _push() -> None:
            await push_docker_image(
                full_image_name,
                on_output=lambda out: _emit("push", _strip_ansi(out)),
            )

        async def _on_push_retry(ctx: RetryContext) -> None:
            _emit(
                "push",
                f"Push failed (attempt {ctx.attempt}), "
                f"retrying in {ctx.delay:.1f}s…",
            )
            if self._retry and self._retry.on_retry:
                result = self._retry.on_retry(ctx)
                if asyncio.iscoroutine(result):
                    await result

        push_retry = RetryConfig(
            max_attempts=self._retry.max_attempts if self._retry else 3,
            should_retry=self._retry.should_retry if self._retry else None,
            on_retry=_on_push_retry,
        )

        await _with_retry("snapshots.pushDockerImage", _push, push_retry)

        return {
            "image": full_image_name,
            "architecture": architecture,
        }
